### DISCUSSION TITLE
Can't get nvim-java to work on custom lazyvim

### DISCUSSION THREAD

**Discussion by Konadu-Akwasi-Akuoko** [nvim-java/nvim-java#283] [Q&A]
Hello all, I just started using neovim, and I wanted to install nvim-java, since I sometimes use Java. But it seems I cannot get it to work (begginer neovimmer here).

So I went to [starter configs recommended for newbies section](https://github.com/nvim-java/nvim-java?tab=readme-ov-file#starter-configs-recommend-for-newbies), and since I am using LazyVim, I headed straight to the [lazyvim starter](https://github.com/nvim-java/starter-lazyvim).

So following the code in the starter lazyvim repo, I added this code to my `nvim-java.lua` file:

```lua
return {
  "nvim-java/nvim-java",
  config = false,
  dependencies = {
    {
      "neovim/nvim-lspconfig",
      opts = {
        servers = {
          jdtls = {
            -- your jdtls configuration goes here
          },
        },
        setup = {
          jdtls = function()
            require("java").setup({
              -- your nvim-java configuration goes here
            })
          end,
        },
      },
    },
  },
}
```

After that I went over to my `treesitter.lua` file, I already had my config there, so I added `java` to the ensure installed option, like so:

```lua
return {
	"nvim-treesitter/nvim-treesitter",
	event = { "BufReadPre", "BufNewFile" },
	build = ":TSUpdate",
	dependencies = {
		"windwp/nvim-ts-autotag",
	},
	config = function()
		-- import nvim-treesitter plugin
		local treesitter = require("nvim-treesitter.configs")

		-- configure treesitter
		treesitter.setup({ -- enable syntax highlighting
			highlight = {
				enable = true,
			},
			-- enable indentation
			indent = { enable = true },

			-- ensure these language parsers are installed
			ensure_installed = {
				"json",
				"javascript",
				"java",
				"typescript",
				"tsx",
				"yaml",
				"html",
				"css",
				"prisma",
				"markdown",
				"markdown_inline",
				"svelte",
				"graphql",
				"bash",
				"lua",
				"vim",
				"dockerfile",
				"gitignore",
				"query",
				"vimdoc",
				"c",
				-- personal, added by Akwasi
				"angular",
				"c_sharp",
				"cpp",
				"elixir",
				"java",
				"markdown",
				"python",
				"rust",
				"scss",
				"sql",
				"toml",
			},
			incremental_selection = {
				enable = true,
				keymaps = {
					init_selection = "<C-space>",
					node_incremental = "<C-space>",
					scope_incremental = false,
					node_decremental = "<bs>",
				},
			},
		})

		require("nvim-ts-autotag").setup({
			opts = {
				-- Defaults
				enable_close = true, -- Auto close tags
				enable_rename = true, -- Auto rename pairs of tags
				enable_close_on_slash = false, -- Auto close on trailing </
				filetypes = {
					"html",
					"javascript",
					"typescript",
					"javascriptreact",
					"typescriptreact",
					"svelte",
					"vue",
					"tsx",
					"jsx",
					"rescript",
					"css",
					"lua",
					"xml",
					"php",
					"markdown",
				},
			},
		})
	end,
}
``` 

And after doing that I had this error:

```txt
Failed to run `config` for mason.nvim

...asi_akuoko/.config/nvim/lua/akwasi/plugins/lsp/mason.lua:12: loop or previous error loading module 'mason-lspconfig'

# stacktrace:
  - lua/akwasi/plugins/lsp/mason.lua:12 _in_ **config**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:1
  - lua/akwasi/plugins/lsp/lspconfig.lua:16 _in_ **config**
  - lua/akwasi/lazy.lua:14
  - init.lua:2
  ```
  
Then I remembered that the [main docs](https://github.com/nvim-java/nvim-java?tab=readme-ov-file#custom-configuration-instructions) said something about setting up before lspconfig, so I went into my mason.lua and added the `require("java").setup` and `require("lspconfig").jdtls.setup({})` to it, so here is my `mason.lua` file:

```lua
return {
	"williamboman/mason.nvim",
	dependencies = {
		"williamboman/mason-lspconfig.nvim",
		"WhoIsSethDaniel/mason-tool-installer.nvim",
	},
	config = function()
		-- import mason
		local mason = require("mason")

		-- import mason-lspconfig
		local mason_lspconfig = require("mason-lspconfig")

		local mason_tool_installer = require("mason-tool-installer")

    -- Setup nvim-java
		require("java").setup()
		require("lspconfig").jdtls.setup({})

		-- enable mason and configure icons
		mason.setup({
			ui = {
				icons = {
					package_installed = "✓",
					package_pending = "➜",
					package_uninstalled = "✗",
				},
			},
		})

		mason_lspconfig.setup({
			-- list of servers for mason to install
			ensure_installed = {
				"angularls@17.3.2",
				"tsserver",
				"html",
				"cssls",
				"tailwindcss",
				"svelte",
				"lua_ls",
				"graphql",
				"emmet_ls",
				"prismals",
				"pyright",
			},
		})

		mason_tool_installer.setup({
			ensure_installed = {
				"prettier", -- prettier formatter
				"stylua", -- lua formatter
				-- "isort", -- python formatter
				-- "black", -- python formatter,
				-- "pylint", -- python linter
				"eslint_d", -- js linter
			},
		})
	end,
}
```

But still I am having this same error:
```
Failed to run `config` for mason.nvim

...asi_akuoko/.config/nvim/lua/akwasi/plugins/lsp/mason.lua:12: loop or previous error loading module 'mason-lspconfig'

# stacktrace:
  - lua/akwasi/plugins/lsp/mason.lua:12 _in_ **config**
  - /mason-lspconfig.nvim/lua/mason-lspconfig/init.lua:1
  - lua/akwasi/plugins/lsp/lspconfig.lua:16 _in_ **config**
  - lua/akwasi/lazy.lua:14
  - init.lua:2
  ```
  
  Please how can I fix this and install nvim-java in peace, thanks for the help in advance. And in case someone want's to take a look at my [config, it's on github](https://github.com/Konadu-Akwasi-Akuoko/dotfiles). I am commenting the `require setup java` steps in my `mason.lua` cause it is crashing things, and also the `nvim-java.lua` file I'll change it to `nvim-java._lua` so that it will not be loaded.

**Comment by s1n7ax** [comment]
You shouldn't call `mason_lspconfig.setup` or other `setup` calls for that matter manually, because LazyVim does that in the background.

If you want to add more language support, for example I see you have angular included, what you should do is add [angular extra](https://www.lazyvim.org/extras/lang/angular). This is how I have added the lazy extras in my person LazyVim config.
https://github.com/s1n7ax/lazyvim-dotnvim/blob/d7f66515a1efc261af817ccee92765ed94cccea4/lua/config/lazy.lua#L28-L32

BUT, lets say you really want to override a configuration, then you should use `opts` property to override it. Here you can see I have added many treesitter parsers without calling the `setup` function.
https://github.com/s1n7ax/lazyvim-dotnvim/blob/d7f66515a1efc261af817ccee92765ed94cccea4/lua/plugins/treesitter/init.lua?plain=1#L3-L38

Now the error you get is due to require `loop`. `A` imports `B` but to import `B` you should have `A` kind of situation. Fix the issues I mentioned previously then this error will magically go away

**Comment by Konadu-Akwasi-Akuoko** [comment]
Thanks a lot for replying, but so I tried adding this code to my [lazy.lua](https://github.com/Konadu-Akwasi-Akuoko/dotfiles/blob/main/lua/akwasi/plugins/lazygit.lua) file:
```lua
require("lazy").setup({
  spec = {
    { "LazyVim/LazyVim", import = "lazyvim.plugins" },
    { import = "lazyvim.plugins.extras.lang.angular" },
    { import = "plugins" },
  },
})
```
Specifically the `{ "LazyVim/LazyVim", import = "lazyvim.plugins" },` and ` { import = "lazyvim.plugins.extras.lang.angular" },` line,  but when I did that it replaced all my configs with one from lazyvim, so I removed this line `{ "LazyVim/LazyVim", import = "lazyvim.plugins" },`, but after removing that I get this error:
```
Error detected while processing /home/akwasi_akuoko/.config/nvim/init.lua:
No specs found for module "lazyvim.plugins.extras.lang.angular"
Press ENTER or type command to continue
```

So it seems I am using Lazy as my extensions manager, but I don't want to use all the plugins that comes with lazyvim?
As of now, I can't even do `:LazyHealth`, cause I don't have the plugins from `LazyVim`, but since I am using `Lazy` as my plugin manager I can do `Lazy` and it shows me all of my plugins.

 I don't know is that why it is not working, I need to install LazyVim before it works?
 
 EDIT:
 
 So following up on your second part of the answer, I have seen that I don't need to even call the `config` function, but using `opts` works, example is my treesitter file, so I am trying to replace all `.setup` in my file, so when I am done I'll let you know, thanks a lot.

**Reply by s1n7ax** [reply]
If you want to disable the plugins you should create a spec and mark it as enabled = false. Of you don't find most of them useful, you are better off with kickstarter config. 

This is how you should disable the unwanted plugins

https://github.com/s1n7ax/lazyvim-dotnvim/blob/main/lua/plugins/disabled-plugins/init.lua

The issue you are getting is unrelated to the plugin. Unfortunately you kinda have to understand how things work to get most of the plugins working
